### PR TITLE
Add "Fill-in-Middle" pipeline

### PR DESCRIPTION
**Pull request by tanaymeh**
# What does this PR do?

This PR adds the Fill-in-Middle pipeline to 🤗 transformers. 

FIM objective was proposed in [Efficient Training of Language Models to Fill in the Middle](https://arxiv.org/abs/2207.14255). They showed that autoregressive language models can learn to infill text after applying a straightforward transformation to the dataset, which simply moves a span of text from the middle of a document to its end.

As discussed in #27059

## Who can review?
@sayakpaul @ArthurZucker

### CODE REVIEW

File: src/transformers/pipelines/fill_in_middle.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+
+class ReturnType(enum.Enum):
+    TENSORS = 0
+    NEW_TEXT = 1
+    FULL_TEXT = 2
+
+
+if is_torch_available():
+    FIM_SUPPORTED_MODELS_TORCH = [
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:16]: yep this sounds alright to me. The mapping could just check if the model name supports it, as some model do / some don't but it could be a bit too restrictive. We can just check if the model type is in the AUTO_MODEL_FOR_CONDITIONAL_GENERATION + if it has the necessary tokens added to the vocab (base the check on models that currently support this + incentives people to use the same code notation for the tokens that serve as prefix / suffix etc

File: src/transformers/pipelines/fill_in_middle.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class FIMPipeline(Pipeline):
+    """
+    Fill-in-the-middle Language generation pipeline. This pipeline predicts the middle part of a text given it's prefix and suffix
+    Example:
+    ```python
+    >>> from transformers import pipeline
+    >>> PROMPT = '''
+        def fibonacci(x: int) -> int:
+            <FILL_ME>
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:35]: down with this as that's what was done with CodeLLama! 
Let's make sure that the pipeline supports returning the filled code with the context in the ocrrect order!

File: src/transformers/pipelines/fill_in_middle.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+        # BS x SL
+        generated_sequence = self.model.generate(
+            input_ids=input_ids, attention_mask=attention_mask, **generate_kwargs
+        )
+        out_b = generated_sequence.shape[0]
+        if self.framework == "pt":
+            generated_sequence = generated_sequence.reshape(
+                in_b, out_b // in_b, *generated_sequence.shape[1:]
+            )
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:281]: we need better variable names overall!

**Review by ArthurZucker — COMMENTED** [review]
Thanks for opening the PR! Aligned with you on using the <FILL_ME> string to split the input. ( by defualt but should be something we can change).  StarCoder does not natively support that so the pipeline could actually be the one doing this. 

Otherwise overall good start!

File: src/transformers/pipelines/fill_in_middle.py (hunk 1 above)
Comment (tanaymeh) [reply src/transformers/pipelines/fill_in_middle.py:16]: Is `AUTO_MODEL_FOR_CONDITIONAL_GENERATION` in `modeling_auto.py`? I am unable to find it.

**Comment by tanaymeh** [comment]
FIMPipeline is now compatible with CodeLlama family of models.

**Comment by ArthurZucker** [comment]
Sure, could you make sure the CIs are green?

**Comment by tanaymeh** [comment]
@ArthurZucker The documentation CI is failing consistently with the following error:

```
ERROR src/transformers/pipelines/fill_in_middle.py - ValueError: line 14 of the docstring for transformers.pipelines.fill_in_middle.FIMPipeline has inconsistent leading whitespace: 'def fib(x: int) -> int:'
```
However, in the example of the FIM pipeline in the docstring, the spaces are a part of the generated code output.

```python
>>> from transformers import pipeline
>>> PROMPT = '''
        def fib(x: int) -> int:
            <FILL_ME>
            return fib(x-1) + fib(x-2)
    '''
>>> generator = pipeline(model="codellama/CodeLlama-7b-hf")
>>> generator(PROMPT, do_sample=False)
[{'generated_text': "\ndef fib(x: int) -> int:\n\tif x == 0:\n\t\treturn 0\n\tif x == 1:\n\t\treturn 1\n\telse:\n\t\treturn fib(x-1) + fib(x-2)\n"}]
```
Should I ignore this?

**Comment by ArthurZucker** [comment]
you can probably ignore it with ` # doctest: +SKIP`

File: src/transformers/pipelines/__init__.py [hunk 4]
```diff
@@ -254,14 +282,35 @@
...
         "pt": (AutoModelForMaskedLM,) if is_torch_available() else (),
-        "default": {"model": {"pt": ("distilroberta-base", "ec58a5b"), "tf": ("distilroberta-base", "ec58a5b")}},
+        "default": {
+            "model": {
+                "pt": ("distilroberta-base", "ec58a5b"),
+                "tf": ("distilroberta-base", "ec58a5b"),
+            }
+        },
+        "type": "text",
+    },
```
Comment (ArthurZucker) [inline src/transformers/pipelines/__init__.py:292]: should not be changed

File: src/transformers/pipelines/text_generation.py
Comment (ArthurZucker) [inline src/transformers/pipelines/text_generation.py:1]: let' s revert unrelated changes

File: src/transformers/pipelines/fill_in_middle.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    "codellama/CodeLlama-13b-hf",
+    "codellama/CodeLlama-34b-hf",
+    "codellama/CodeLlama-7b-Python-hf",
+    "codellama/CodeLlama-13b-Python-hf",
+    "codellama/CodeLlama-34b-Python-hf",
+]
+
+
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class FIMPipeline(Pipeline):
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:35]: let's use camel casing
```suggestion
class FimPipeline(Pipeline):
```

File: src/transformers/pipelines/fill_in_middle.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    """
+    Fill-in-the-middle Language generation pipeline. This pipeline predicts the middle part of a text given it's prefix and suffix
+    Example:
+    ```python
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:39]:
```suggestion

    Example:
    ```python
```

File: src/transformers/pipelines/fill_in_middle.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    >>> PROMPT = '''
+        def fib(x: int) -> int:
+            <FILL_ME>
+            return fib(x-1) + fib(x-2)
+    '''
+    >>> generator = pipeline(model="codellama/CodeLlama-7b-hf")
+    >>> print(generator(PROMPT, do_sample=False))
+    [{'generated_text': "\ndef fib(x: int) -> int:\n\tif x == 0:\n\t\treturn 0\n\tif x == 1:\n\t\treturn 1\n\telse:\n\t\treturn fib(x-1) + fib(x-2)\n"}]
+    ```
+    Learn more about the basics of using a pipeline in the [pipeline tutorial](../pipeline_tutorial). You can pass text
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:50]:
```suggestion
    ```

    Learn more about the basics of using a pipeline in the [pipeline tutorial](../pipeline_tutorial). You can pass text
```

this should help fix the failing ci

File: src/transformers/pipelines/fill_in_middle.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    text generation parameters in [Text generation strategies](../generation_strategies) and [Text
+    generation](text_generation).
+    This language generation pipeline can currently be loaded from [`pipeline`] using the following task identifier:
+    `"fill-in-middle"`.
+    The models that this pipeline can use are models that have been trained with an autoregressive language modeling
+    objective, which includes the uni-directional models in the library (e.g. gpt2). See the list of available models
+    on [huggingface.co/models](https://huggingface.co/models?filter=text-generation).
+    """
+
+    DEFAULT_MODE = "psm"
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:61]: that' s not very explicit so let' s use better name

File: src/transformers/pipelines/fill_in_middle.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+            raise ValueError(f"The model '{self.model.name_or_path}' is not supported for the Fill-in-Middle task.")
+
+        self.check_model_type(
+            TF_MODEL_FOR_CAUSAL_LM_MAPPING_NAMES if self.framework == "tf" else MODEL_FOR_CAUSAL_LM_MAPPING_NAMES
+        )
+        if "prefix" not in self._preprocess_params:
+            # This is very specific. The logic is quite complex and needs to be done
+            # as a "default".
+            # It also defines both some preprocess_kwargs and generate_kwargs
+            # which is why we cannot put them in their respective methods.
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:80]:
```suggestion
```

File: src/transformers/pipelines/fill_in_middle.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+        add_special_tokens=False,
+        **generate_kwargs,
+    ):
+        preprocess_params = {"add_special_tokens": add_special_tokens}
+        preprocess_params["infill_token"] = infill_token or self.DEFAULT_INFILL_TOKEN
+
+        if mode is not None:
+            preprocess_params["mode"] = mode.lower() if mode.lower() in ["psm", "spm"] else self.DEFAULT_MODE
+        else:
+            preprocess_params["mode"] = self.DEFAULT_MODE
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:129]: i have no idea what psm or spm mean. Let's document it!

File: src/transformers/pipelines/fill_in_middle.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+
+        if prefix is not None:
+            preprocess_params["prefix"] = prefix
+        if prefix:
+            prefix_inputs = self.tokenizer(
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:134]: I think we should encode both inputs at the same time to not have to merge the attention masks

File: src/transformers/pipelines/fill_in_middle.py [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+            prefix (`str`, *optional*):
+                Prefix added to prompt.
+            handle_long_generation (`str`, *optional*):
+                By default, this pipelines does not handle long generation (ones that exceed in one form or the other
+                the model maximum length). There is no perfect way to adress this (more info
+                :https://github.com/huggingface/transformers/issues/14033#issuecomment-948385227). This provides common
+                strategies to work around that problem depending on your use case.
+
+                - `None` : default strategy where nothing in particular happens
+                - `"hole"`: Truncates left of input, and leaves a gap wide enough to let generation happen (might
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:317]: I don't think we should do that in the pipeline it' s not specific to FIM

File: src/transformers/pipelines/fill_in_middle.py [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+                    + self.DEFAULT_MIDDLE_TOKEN
+                )
+            else:
+                prompt_text = (
+                    self.DEFAULT_SUFFIX_TOKEN
+                    + input_suffix
+                    + self.DEFAULT_PREFIX_TOKEN
+                    + input_prefix
+                    + self.DEFAULT_MIDDLE_TOKEN
+                )
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:215]: not sure I understand why we only support Llama for this case?

**Review by ArthurZucker — COMMENTED** [review]
Thanks for working on this 🤗

there seems to be changes in the __init__ that are unrelated to the PR so we should revert them! 

Also let' s not add extra complexity, and keep it simple. 2 modes, and support outputing the formated text with the code inside!

**Comment by ArthurZucker** [comment]
Also cc @Rocketknight1

File: src/transformers/pipelines/fill_in_middle.py (hunk 13 above)
Comment (tanaymeh) [reply src/transformers/pipelines/fill_in_middle.py:215]: Of the two models (that I am aware of) that support the FIM objective, CodeLlama has the support of directly using the `<FILL_ME>` token whereas starcoder doesn't. We have to manually add the prefix, suffix and middle tokens, hence the check.

Comment (ArthurZucker) [reply src/transformers/pipelines/fill_in_middle.py:215]: Yes, but unifying this would be better! But got it for this now worries

**Comment by ArthurZucker** [comment]
No worries, ping me whenever for another review!

**Comment by tanaymeh** [comment]
@ArthurZucker About your suggestion on merging the CodeLlama FIM logic with other model variants: It would complicate things a lot in the pipeline code as it will require a lot of code from the CodeLlama tokenizer to be included directly in the pipeline code since the tokenizer deals with much of the FIM related operations in the `tokenize()` and other functions.

Keeping them separate (i.e; For models that support a `<FILL_ME>` token out of the box vs. those that don't support it like starcoder) will be easier to debug and maintain and will remove redundancy.

FYI, I am thinking of something like this:

Supported model list will have all model names, corresponding FIM tokens and if they support Out of the box Infilling token or not (if they don't, it will be None in place of infill token)
```python
SUPPORTED_MODELS = {
    "bigcode/starcoder": ("<fim_prefix>", "<fim_middle>", "<fim_suffix>", None),
    "codellama/CodeLlama-7b-hf": ("▁<PRE>", "▁<MID>", "▁<SUF>", "<FILL_ME>"),
    # other models
}

...

infill_token = self.SUPPORTED_MODELS[self.model.name_or_path]

if not infill_token:
    # Extract prefix and suffix
    input_prefix, input_suffix = self.extract_prefix_suffix(prompt_text, infill_token)

    if mode == "psm":
        prompt_text = (
            self.DEFAULT_PREFIX_TOKEN
            + input_prefix
            + self.DEFAULT_SUFFIX_TOKEN
            + input_suffix
            + self.DEFAULT_MIDDLE_TOKEN
        )
    else:
        prompt_text = (
            self.DEFAULT_SUFFIX_TOKEN
            + input_suffix
            + self.DEFAULT_PREFIX_TOKEN
            + input_prefix
            + self.DEFAULT_MIDDLE_TOKEN
        )
else:
    # Tokenizer directly, since the Infill token is supported out of the box and doesn't need any token re-arrangement
    ...
```

Please let me know your opinion on this!

**Comment by ArthurZucker** [comment]
IMO that is exactly the purpose of this pipeline. The functions should not necessarily have been part of the tokenizer as they are only need for the FIM task. So I am more saying let's not rely on any tokenizer, but handle the processing in the pipeline

**Comment by tanaymeh** [comment]
> IMO that is exactly the purpose of this pipeline. The functions should not necessarily have been part of the tokenizer as they are only need for the FIM task. So I am more saying let's not rely on any tokenizer, but handle the processing in the pipeline

I see, thanks for clarifying, will push the updates soon!

File: src/transformers/pipelines/__init__.py [hunk 14]
```diff
@@ -257,6 +258,15 @@
         "default": {"model": {"pt": ("distilroberta-base", "ec58a5b"), "tf": ("distilroberta-base", "ec58a5b")}},
         "type": "text",
     },
+    "fill-in-middle": {
+        "impl": FimPipeline,
+        "tf": (TFAutoModelForMaskedLM,) if is_tf_available() else (),
```
Comment (ArthurZucker) [inline src/transformers/pipelines/__init__.py:263]:
```suggestion

```

File: src/transformers/pipelines/__init__.py [hunk 15]
```diff
@@ -257,6 +258,15 @@
...
+        "pt": (AutoModelForMaskedLM,) if is_torch_available() else (),
+        "default": {
+            "model": {"pt": ("codellama/CodeLlama-7b-hf", "ec58a5b"), "tf": ("codellama/CodeLlama-7b-hf", "ec58a5b")}
```
Comment (ArthurZucker) [inline src/transformers/pipelines/__init__.py:266]:
```suggestion
            "model": {"pt": ("codellama/CodeLlama-7b-hf", "ec58a5b")}
```

File: src/transformers/pipelines/fill_in_middle.py [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+    The models that this pipeline can use are models that have been trained with an autoregressive language modeling
+    objective, which includes the uni-directional models in the library (e.g. gpt2). See the list of available models
+    on [huggingface.co/models](https://huggingface.co/models?filter=text-generation).
+    """
+
+    # There are two modes in which infilling is supported: PSM (Prefix-Suffix-Middle) and SPM (Suffix-Prefix-Middle)
+    # CodeLlama's HF implementation has a third mode called 'suffix-first' which is added here for total compatibility with CodeLlama family
+    # The placement of prefix and suffix along with their respective sentinel tokens depends on the mode
+    # More information on this: https://arxiv.org/abs/2207.14255
+    DEFAULT_INFILL_MODE = "psm"
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:67]: again let's use full names please

File: src/transformers/pipelines/fill_in_middle.py [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+    DEFAULT_INFILL_TOKEN = "<FILL_ME>"
+    ALL_INFILL_MODES = ["psm", "spm", "suffix-first"]
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:69]: same here

File: src/transformers/pipelines/fill_in_middle.py [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Check if the provided model is among the supported ones
+        if self.model.name_or_path not in SUPPORTED_MODELS:
+            raise ValueError(f"The model '{self.model.name_or_path}' is not supported for the Fill-in-Middle task.")
+
+        # Get the Prefix, Suffix and Middle tokens for the corresponding model
+        self.PREFIX_TOKEN, self.MIDDLE_TOKEN, self.SUFFIX_TOKEN = SUPPORTED_MODELS[self.model.name_or_path]
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:78]: that is not super super scalable, if someone has a custome model he also needs to register these variables. We could add them to the generation config IMO

File: src/transformers/pipelines/fill_in_middle.py [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+        """
+        if infill_token not in prompt_text:
+            raise ValueError(f"Infill token '{infill_token}' not in the prompt")
+
+    def extract_prefix_suffix(self, input_text, infill_token):
+        """
+        Extracts the prefix and suffix from the input text.
+        """
+        prefix, _, suffix = input_text.partition(infill_token)
+        return prefix, suffix
```
Comment (ArthurZucker) [inline src/transformers/pipelines/fill_in_middle.py:113]: we probably don't need that one liner

**Review by ArthurZucker — COMMENTED** [review]
Alright that looks a lot better. Let's add tests now!

File: src/transformers/pipelines/fill_in_middle.py (hunk 19 above)
Comment (tanaymeh) [reply src/transformers/pipelines/fill_in_middle.py:113]: Are you suggesting that instead of making a function for extracting prefix and suffix, I just do it in-line? Because we need this functionality to extract the prefix and suffix from the user's provided prompt.

**Comment by tanaymeh** [comment]
@ArthurZucker I added the naming and init file changes as you requested. Working on the tests by taking the text-generation tests as a reference.

One doubt: The smallest model that supports FIM (that I know of) is `codellama-7b` which is still a pretty huge model and should use GPU for testing. Is there a way to do the tests without using such a big model?

File: src/transformers/pipelines/fill_in_middle.py (hunk 19 above)
Comment (ArthurZucker) [reply src/transformers/pipelines/fill_in_middle.py:113]: yes 👍🏻

**Comment by ArthurZucker** [comment]
You should be able to use 4bit quantization!

File: src/transformers/pipelines/fill_in_middle.py (hunk 18 above)
Comment (tanaymeh) [reply src/transformers/pipelines/fill_in_middle.py:78]: @ArthurZucker For this to work, we need to first add the infill tokens to GenerationConfigs for all Codellama models and the Starcoder model right?

Comment (ArthurZucker) [reply src/transformers/pipelines/fill_in_middle.py:78]: Mmm they should be in the tokenizer_config.json.
We can just save the tokens by doing `CodeLlamaTokenizer.from_pretrained("...", prefix_token = ""). It will be saved. If they are not in the tokenizer then error out.

**Comment by tanaymeh** [comment]
> Sounds good. No need for the generation config update . Tokens are string so should be saved in the tokenizer_config.json IMO

@ArthurZucker But for that, we need to save them to the tokenizer_config.json first for the tests to pass right?

**Comment by ArthurZucker** [comment]
Yes, you can probably open a PR to the models and use the `revision`! WDYT?

**Comment by tanaymeh** [comment]
Hi @ArthurZucker, I was out during easter and will work on this now. Sorry for the delay!
